driver/ni/digital_write.cpp

```diff
@@ -74,267 +73,397 @@ ni::DigitalWriteSink::DigitalWriteSink(
+        // the order the keys were pushed into the vector is the order the data is written
         // first see if the key is in the drive_cmd_channel_keys
-        auto it = std::find(this->writer_config.drive_cmd_channel_keys.begin(), this->writer_config.drive_cmd_channel_keys.end(), key);
-        if (it != this->writer_config.drive_cmd_channel_keys.end()){
-            // if so, now find which index it is in the vector (i.e. which channel it is in the writeBuffer)
-            cmd_channel_index = std::distance(this->writer_config.drive_cmd_channel_keys.begin(), it); // this corressponds to where in the order its NI channel was created
+        auto it = std::find(this->writer_config.drive_cmd_channel_keys.begin(),
+                            this->writer_config.drive_cmd_channel_keys.end(), key);
+        if (it != this->writer_config.drive_cmd_channel_keys.end()) {
+            // if so, now find which index it is in the vector (i.e. which channel it is in the write_buffer)
+            cmd_channel_index = std::distance(
+                this->writer_config.drive_cmd_channel_keys.begin(),
+                it);
+            // this corressponds to where in the order its NI channel was created
             // now we grab the level we'd like to write and put it into that location in the write_buffer
             auto series = frame.series->at(frame_index).uint8();
-            writeBuffer[cmd_channel_index] = series[0];
-            this->writer_config.modified_state_keys.push(this->writer_config.drive_state_channel_keys[cmd_channel_index]);
+            write_buffer[cmd_channel_index] = series[0];
+            this->writer_config.modified_state_keys.push(
+                this->writer_config.state_channel_keys[cmd_channel_index]);
             this->writer_config.modified_state_values.push(series[0]);
         }
         frame_index++;
     }
     return freighter::NIL;
 }
 
+ni::DigitalWriteSink::~DigitalWriteSink() {
+    this->clear_task();
+    if (this->write_buffer) delete[] this->write_buffer;
+}
 
-int ni::DigitalWriteSink::checkNIError(int32 error){
-    if (error < 0){
+void ni::DigitalWriteSink::clear_task() {
+    if (this->check_ni_error(ni::NiDAQmxInterface::ClearTask(task_handle)))
+        this->log_error(
+            "failed to clear writer for task " + this->writer_config.task_name);
+}
+
+std::vector<synnax::ChannelKey> ni::DigitalWriteSink::get_cmd_channel_keys() {
+    std::vector<synnax::ChannelKey> keys;
+    for (auto &channel: this->writer_config.channels)
+        if (channel.channel_type != "index") keys.push_back(channel.channel_key);
+    return keys;
+}
+
+std::vector<synnax::ChannelKey> ni::DigitalWriteSink::get_state_channel_keys() {
+    std::vector<synnax::ChannelKey> keys = this->writer_config.state_channel_keys;
+    keys.push_back(this->writer_config.state_index_key);
+    return keys;
+}
+
+int ni::DigitalWriteSink::check_ni_error(int32 error) {
+    if (error < 0) {
```

Comment:
i don't think its worth it for just this one function esp since it needs access to member variables so it would mean coupling the ni sinks and sources with this which is not worth the reduced code duplications